### DISCUSSION TITLE
Gruntwork Pipelines using wrong terraform version

### DISCUSSION THREAD

**Discussion by jeffreymlewis** [gruntwork-io/knowledge-base#789] [How do I] [answered]
I've finally got Gruntwork Pipelines running. However, it's using the wrong version of terraform (see below). What is the correct way to use a specific terraform version in Pipelines? I tried using the `.gruntwork/config.yml` file in infra-live, resulting in the output below.

(.gruntwork/config.yml)
```
pipelines:
  cli-version: v0.2.4
  terraform-version: 1.5.5
  terragrunt-version: 0.50.14
```

```
Installing Terraform v1.5.5
Downloading release tarball from https://releases.hashicorp.com/terraform/1.5.5/terraform_1.5.5_linux_amd64.zip
#=#=#                                                                         

                                                                           0.0%
                                                                           0.1%
#                                                                          2.5%
###                                                                        5.1%
###################                                                       27.7%
###########################                                               37.7%
###################################                                       49.9%
######################################                                    53.9%
#############################################                             63.6%
#####################################################                     74.1%
#############################################################             85.1%
######################################################################## 100.0%
Downloading SHA hash file from https://releases.hashicorp.com/terraform/1.5.5/terraform_1.5.5_SHA256SUMS
Not instructed to use Local PGP (/home/runner/.tfenv/use-{gpgv,gnupg}) & No keybase install found, skipping OpenPGP signature verification
Archive:  /tmp/tfenv_download.u8wWoG/terraform_1.5.5_linux_amd64.zip
  inflating: /home/runner/.tfenv/versions/1.5.5/terraform  
Installation of terraform v1.5.5 successful. To make this your default version, run 'tfenv use 1.5.5'
Switching default version to v1.5.5
Default version (when not overridden by .terraform-version or TFENV_TERRAFORM_VERSION) is now: 1.5.5
2023/12/05 01:21:36 Creating directory for terragrunt: /home/runner/.terragrunt.versions/
Downloading https://github.com/gruntwork-io/terragrunt/releases/download/v0.50.14/terragrunt_linux_amd64 to terragrunt_linux_amd64
Downloading ...
47453543 bytes downloaded.
Switched terragrunt to version "0.50.14" 
Run dsaltares/fetch-gh-release-asset@1.1.1
Run sudo mv /tmp/pipelines /usr/local/bin/pipelines
Run echo "MACHINE_USER_NAME=$(gh api /user | jq .login)" >> "$GITHUB_OUTPUT"
Run # CD Infra-live directory && Run Pipelines-execute
time="2023-12-05T01:21:39Z" level=info msg="This program sends telemetry to Gruntwork. To disable, set PIPELINES_TELEMETRY_OPT_OUT=true as an environment variable" binary=pipelines version=v0.2.4
time="2023-12-05T01:21:39Z" level=info msg="Configuring git to use HTTPS with the provided machine user token and name" binary=pipelines version=v0.2.4
time="2023-12-05T01:21:39Z" level=info msg="Starting Pipelines executor..." binary=pipelines version=v0.2.4
time="2023-12-05T01:21:39Z" level=info msg="Checking out branch: 5e2dec8897f7450c1877c997072fe797f0e6c4fe" binary=pipelines version=v0.2.4
HEAD is now at 5e2dec8 Add `.terraform-version` file
time="2023-12-05T01:21:39Z" level=info msg="Current working directory: /home/runner/work/infrastructure-pipelines/infrastructure-pipelines/infrastructure-live" binary=pipelines version=v0.2.4
time="2023-12-05T01:21:39Z" level=info msg="Changed into directory: dev/us-east-2/dev/data-stores/s3/hrg-dev-backup-archives" binary=pipelines version=v0.2.4
time="2023-12-05T01:21:39Z" level=info msg="Executing Terragrunt with args: plan" binary=pipelines version=v0.2.4

Initializing the backend...

Successfully configured the backend "s3"! Terraform will automatically
use this backend unless the backend configuration changes.
Initializing modules...
Downloading git::ssh://git@github.com/gruntwork-io/terraform-aws-security.git?ref=v0.69.0 for s3_bucket_logs...
- s3_bucket_logs in .terraform/modules/s3_bucket_logs/modules/private-s3-bucket
Downloading git::ssh://git@github.com/gruntwork-io/terraform-aws-security.git?ref=v0.69.0 for s3_bucket_primary...
- s3_bucket_primary in .terraform/modules/s3_bucket_primary/modules/private-s3-bucket
Downloading git::ssh://git@github.com/gruntwork-io/terraform-aws-security.git?ref=v0.69.0 for s3_bucket_replica...
- s3_bucket_replica in .terraform/modules/s3_bucket_replica/modules/private-s3-bucket
╷
│ Error: Unsupported Terraform Core version
│ 
│   on provider_version_override.tf line 3, in terraform:
│    3:   required_version = "<= 1.5.5"
│ 
│ This configuration does not support Terraform version 1.6.4. To proceed,
│ either choose another supported Terraform version or update this version
│ constraint. Version constraints are normally set for good reason, so
│ updating the constraint may lead to other errors or unexpected behavior.
╵
```

---

<ins datetime="2023-12-05T01:37:24Z">
  <p><a href="https://support.gruntwork.io/hc/requests/110616">Tracked in ticket #110616</a></p>
</ins>

**Comment by gitsstewart** [comment]
@jeffreymlewis our new Pipelines product does not yet provide support for Terraform beyond version 1.5.5 due to the changes in Hashicorp licensing.

**Reply by oredavids** [reply]
@jeffreymlewis, the config you referenced in your infra-live is the correct way to specify the Terraform version to use. I see that the specified version of 1.5.5, in the config, is also what is shown in your logs as being installed:

```
Installation of terraform v1.5.5 successful. To make this your default version, run 'tfenv use 1.5.5'
Switching default version to v1.5.5
Default version (when not overridden by .terraform-version or TFENV_TERRAFORM_VERSION) is now: 1.5.5
```

so that is what Pipelines will use except you have a different configuration in your actual code. Our modules restrict using any Terraform version above 1.5.5 like @gitsstewart mentioned.

Do you have a `.terraform-version` included in your git repository in the folder containing the terragrunt file that pipelines is processing?
Also, a snippet of the infra-code you are trying to deploy, including the module source URL, would help to triage. Are you following our [S3 example for pipelines](https://docs.gruntwork.io/pipelines/hello-world/#adding-a-new-s3-bucket)?

**Reply by jeffreymlewis** [reply]
@gitsstewart said
> our new Pipelines product does not yet provide support for Terraform beyond version 1.5.5 ...

Then why is it trying to use terraform v1.6.4 ?

@oredavids said
> ... except you have a different configuration in your actual code.

What are you referring to here? There is no configuration in my "actual code" specifying the version of terraform that Gruntwork Pipelines is using.

> Do you have a .terraform-version included in your git repository ...

No, this file does not exist anywhere in infra-live.

**Reply by jeffreymlewis** [reply]
I enabled debugging via ssh for the `pipelines-execute` step. Here's the state of the runner executing via Gruntwork Pipelines.

```
runner@fv-az790-651:~$ cat ~/.tfenv/version
1.5.5

runner@fv-az790-651:~$ terraform version
Terraform v1.6.5
on linux_amd64
```

It looks like you're trying to use `tfenv` to manage terraform versions on this runner, but it's not working. I've never used tfenv, but I'm guessing it's been installed incorrectly. I'll troubleshoot later when I have some time.

**Reply by oredavids** [reply]
Thanks for spotting that @jeffreymlewis. If we get to a fix first I'll update you.

**Reply by jeffreymlewis** [reply]
HI @oredavids thanks for the quick reply.

The issue is the system-installed version of terraform (/usr/local/bin/terraform) is being found first because of the way `pipelines-execute` is [setting the PATH](https://github.com/gruntwork-io/pipelines-execute/blob/main/action.yml#L78). You either need to remove /usr/local/bin/terraform from the system, or move `$HOME/.tfenv/bin` to the front of the PATH.

Also, I might suggest writing `export PATH=...` into `~/.bash_profile` rather than just exporting as you're doing now (see above). That way, people debugging via ssh will have the same $PATH as the runner.

I also suggest updating your templates and/or bootstrapping code, to add `with_ssh_enabled` as an input for the `Run terragrunt ${{ inputs.terragrunt_command }}` step in infrastructure-pipelines, so that users can enable ssh from infra-live on a case-by-case basis.

**Reply by jeffreymlewis** [reply]
Hmm... looks like you'd also need to update the `pipelines` binary to support calling Run terragrunt with `with_ssh_enabled`. Unfortunately the source for this binary is not available so I can't submit a PR. 👎 

https://github.com/gruntwork-io/pipelines-cli

**Reply by oredavids** [reply]
Thanks @jeffreymlewis. Your summary on the wrong version issue is what I thought may be happening. We'll be working on fixing this today and also incorporate your suggestion on using bash profile for the PATH update. 

cc @yhakbar

**Reply by yhakbar** [reply]
@jeffreymlewis 

The most recent release of the action should address the issues that you're experiencing:
https://github.com/gruntwork-io/pipelines-execute/releases/tag/v2.1.0

Please confirm when convenient, and let us know if you have any other issues!

**Reply by jeffreymlewis** [reply]
@oredavids @yhakbar Still not working, same issue. Did you folks test this? (I'm curious why this is working for you, but not working for me?)

I'll fix this, and submit a PR for you.

Also FYI, putting a `PATH` declaration into `~/.bash_profile` did not have the desired affect, because the `/etc/profile` on this system doesn't source `~/.bash_profile`. So you should probably either remove that, or fix.

```
│ Error: Unsupported Terraform Core version
│ 
│   on provider_version_override.tf line 3, in terraform:
│    3:   required_version = "<= 1.5.5"
│ 
│ This configuration does not support Terraform version 1.6.5. To proceed,
│ either choose another supported Terraform version or update this version
│ constraint. Version constraints are normally set for good reason, so
│ updating the constraint may lead to other errors or unexpected behavior.
```

**Reply by jeffreymlewis** [reply]
@oredavids @yhakbar I've created this PR for you. The thing to know, is that environment variables set in a step, are NOT available in subsequent steps. If you want to do this, you need to use [GITHUB_ENV](https://docs.github.com/en/github-ae@latest/actions/using-workflows/workflow-commands-for-github-actions#setting-an-environment-variable) or [outputs](https://docs.github.com/en/github-ae@latest/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idoutputs).

https://github.com/gruntwork-io/pipelines-execute/pull/8

The above `pipelines-execute` is now working for me. Let me know when you've merged a long-term fix, so I can switch back to using the GW repo (rather then my forked repo). Thanks!

**Reply by yhakbar** [reply, marked as answer]
Hey @jeffreymlewis , thanks for your PR. We've taken a slightly different tactic in this PR:
https://github.com/gruntwork-io/pipelines-execute/pull/9

Do you mind confirming that as of the latest release the action is working in your environment?

Apologies for the delay in resolution.

**Reply by jeffreymlewis** [reply]
It looks like release `v2.1.1` in `pipelines-execute` includes this fix. I'll check it out today.

**Reply by jeffreymlewis** [reply]
The new release works as expected. Thanks @yhakbar !
